### PR TITLE
[Semantic API] Consider on-fail clause for relevent statements in all the node visitors

### LINKED ISSUE #35060: Consider on-fail clause for while statement in all the node visitors
We haven't seem to consider on-fail clause for while statement in the visitors. Need to add that. #34857 is an issue due to this.

### PR DESCRIPTION
**Pull request by DamithaSenevirathne**
## Purpose

Fixes #35060 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/symbols/SymbolsInOnFailClauseTest.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import static io.ballerina.compiler.api.symbols.TypeDescKind.ERROR;
+import static io.ballerina.semantic.api.test.util.SemanticAPITestUtils.getDefaultModulesSemanticModel;
+import static io.ballerina.semantic.api.test.util.SemanticAPITestUtils.getDocumentForSingleSource;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertTrue;
+
+/**
+ * Test cases for on-fail clause.
+ *
+ * @since 2.0.0
```
Comment (pubudu91) [inline tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/symbols/SymbolsInOnFailClauseTest.java:44]: Now that we have a versioning scheme, let's stick to that. Or let's just get rid of this entirely

File: tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/SymbolLookupTest.java [hunk 2]
```diff
@@ -358,6 +358,29 @@ public void testSymbolLookupInFields(int line, int column, int expSymbols, List<
...
+        BLangPackage pkg = packageCompilation.defaultModuleBLangPackage();
+        ModuleID moduleID = new BallerinaModuleID(pkg.packageID);
+
+        List<Symbol> allInScopeSymbols = model.visibleSymbols(srcFile, LinePosition.from(25, 23));
+        List<Symbol> symbols = new ArrayList<>();
+        for (Symbol symbol : allInScopeSymbols) {
+            if (symbol.getModule().get().id().equals(moduleID) && symbol.kind() == SymbolKind.VARIABLE) {
+                symbols.add(symbol);
+            }
+        }
```
Comment (dulajdilshan) [inline tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/SymbolLookupTest.java:379]: `SemanticAPITestUtils` has a function `getSymbolsInFile()`. We can get the symbols filtered. shall we use this?

File: tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/SymbolLookupTest.java [hunk 3]
```diff
@@ -358,6 +358,29 @@ public void testSymbolLookupInFields(int line, int column, int expSymbols, List<
...
+
+        assertEquals(symbols.size(), 2);
```
Comment (dulajdilshan) [inline tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/SymbolLookupTest.java:381]: Don't we need to assert symbol informations for this case?

File: tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/allreferences/FindRefsInFailTrtryStmtTest.java [hunk 4]
```diff
@@ -64,6 +64,12 @@
                 {87, 21, location(87, 20, 22),
                         List.of(location(87, 20, 22),
                                 location(88, 20, 22))
+                },
+                {104, 20, location(103, 20, 21),
+                        List.of(location(103, 20, 21),
+                                location(104, 20, 21)
+                        )
+
```
Comment (dulajdilshan) [inline tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/allreferences/FindRefsInFailTrtryStmtTest.java:72]: Do we need this extra line?

File: tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/symbols/SymbolsInOnFailClauseTest.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+    @BeforeClass
+    public void setup() {
+        Project project = BCompileUtil.loadProject("test-src/symbols/symbols_in_on_fail_test.bal");
+        model = getDefaultModulesSemanticModel(project);
+        srcFile = getDocumentForSingleSource(project);
+    }
+
+    @Test(dataProvider = "OnFailPosProvider")
+    public void onFailTest(int line, int sCol, int eCol, TypeDescKind typeKind, String name) {
```
Comment (dulajdilshan) [inline tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/symbols/SymbolsInOnFailClauseTest.java:59]: Shall we rename this as `testOnFailCluase` or `testOnFail`?. The name of a method of a class must be in the "do-something" pattern.
Ex: 
- checknode()
- analyzeFunction()
- visitFunction()

File: tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/symbols/SymbolsInOnFailClauseTest.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        Optional<TypeSymbol> type = model.typeOf(
+                LineRange.from(srcFile.name(), LinePosition.from(line, sCol), LinePosition.from(line, eCol)));
+        assertTrue(type.isPresent());
+        assertEquals(type.get().typeKind(), typeKind);
+        type.get().getName().ifPresent(tName -> assertEquals(tName, name));
+    }
+
+    @DataProvider(name = "OnFailPosProvider")
+    public Object[][] onFailTest() {
```
Comment (dulajdilshan) [inline tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/symbols/SymbolsInOnFailClauseTest.java:68]: I would suggest `getOnFailSymbolPositions` as the methhod name

File: tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/symbols/SymbolsInOnFailClauseTest.java (hunk 1 above)
Comment (DamithaSenevirathne) [reply tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/symbols/SymbolsInOnFailClauseTest.java]: Removed in 0c8d1ed8a6d5d93eacc14e6b7d5ebafb004dfe49

File: tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/allreferences/FindRefsInFailTrtryStmtTest.java (hunk 4 above)
Comment (DamithaSenevirathne) [reply tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/allreferences/FindRefsInFailTrtryStmtTest.java]: Removed in 0c8d1ed8a6d5d93eacc14e6b7d5ebafb004dfe49

File: tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/symbols/SymbolsInOnFailClauseTest.java (hunk 5 above)
Comment (DamithaSenevirathne) [reply tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/symbols/SymbolsInOnFailClauseTest.java]: Updated in 0c8d1ed8a6d5d93eacc14e6b7d5ebafb004dfe49

File: tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/symbols/SymbolsInOnFailClauseTest.java (hunk 6 above)
Comment (DamithaSenevirathne) [reply tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/symbols/SymbolsInOnFailClauseTest.java]: Updated in 0c8d1ed8a6d5d93eacc14e6b7d5ebafb004dfe49

File: tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/SymbolLookupTest.java (hunk 2 above)
Comment (DamithaSenevirathne) [reply tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/SymbolLookupTest.java]: Updated in d8a85d3cf66bd59ad3827b8092e1dbe7c59bb0d7

File: tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/SymbolLookupTest.java (hunk 3 above)
Comment (DamithaSenevirathne) [reply tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/SymbolLookupTest.java]: Updated in d8a85d3cf66bd59ad3827b8092e1dbe7c59bb0d7

File: tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/SymbolLookupTest.java [hunk 7]
```diff
@@ -358,6 +358,26 @@ public void testSymbolLookupInFields(int line, int column, int expSymbols, List<
...
+
+        BLangPackage pkg = packageCompilation.defaultModuleBLangPackage();
+        ModuleID moduleID = new BallerinaModuleID(pkg.packageID);
+
+        Map<String, Symbol> symbolsInFile = getSymbolsInFile(model, srcFile, 25, 23, moduleID);
+        assertEquals(symbolsInFile.size(), 4);
+        assertEquals(symbolsInFile.get("err").kind(), SymbolKind.VARIABLE);
+        assertEquals(symbolsInFile.get("test").kind(), SymbolKind.FUNCTION);
+        assertEquals(symbolsInFile.get("v").kind(), SymbolKind.PARAMETER);
+        assertEquals(symbolsInFile.get("errRef").kind(), SymbolKind.VARIABLE);
```
[the remark this answers is not in the document]
Comment (DamithaSenevirathne) [reply tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/SymbolLookupTest.java]: Updated in 2b85e7097badd80947edb3afa4db80261093fd91

File: tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/SymbolLookupTest.java [hunk 8]
```diff
@@ -358,6 +358,34 @@ public void testSymbolLookupInFields(int line, int column, int expSymbols, List<
...
+        }
+    }
+
+    @DataProvider(name = "OnFailSymbolPosProvider")
+    public Object[][] getOnFailSymbolPositions() {
+        List<String> expSymbolNames = List.of("err", "test", "v", "errRef");
+        return new Object[][]{
+                {25, 23, 4, expSymbolNames}
+        };
+    }
```
Comment (pubudu91) [inline tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/SymbolLookupTest.java:387]: Are you planning on adding any more test cases to this? If not, no need to use a data provider.

File: tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/symbols/SymbolsInOnFailClauseTest.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    @DataProvider(name = "OnFailPosProvider")
+    public Object[][] getOnFailSymbolPositions() {
+        return new Object[][]{
+                {21, 20, 23, ERROR, "err"},
+                {30, 20, 23, ERROR, "err"},
+                {38, 20, 23, ERROR, "err"},
+                {54, 20, 21, ERROR, "e"},
+                {67, 23, 26, ERROR, "err"},
+                {79, 19, 20, ERROR, "e"},
+                {88, 19, 20, ERROR, "e"}
```
Comment (pubudu91) [inline tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/symbols/SymbolsInOnFailClauseTest.java:74]: If there aren't going to be any different type kinds, we can just hard code the typedesc kind.

Comment (DamithaSenevirathne) [reply tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/symbols/SymbolsInOnFailClauseTest.java]: Updated in 542b249fe324c673d0590e2251dbc794055a345e

File: tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/SymbolLookupTest.java (hunk 8 above)
Comment (DamithaSenevirathne) [reply tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/SymbolLookupTest.java]: Updated in b82a084b8bb603675a2afe7a0d16ca65a086b214

**Review by dulajdilshan — APPROVED** [review]
LGTM

**Comment by dulajdilshan** [comment]
Shall we merge the latest `2201.0.x`

**Review by pubudu91 — APPROVED** [review]

**Pull request merged**